storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3577,6 +3579,238 @@ dberr_t fsp_sys_tablespace_validate()
 }
 #endif /* UNIV_DEBUG */
 
+/** Store the inode information which basically stores
+the page and offset */
+struct inodeInfo
+{
+private:
+  std::map<uint32_t, std::set<uint16_t>> m_used_inodes;
+public:
+  void insert(uint32_t page, uint16_t offset)
+  {
+    m_used_inodes[page].insert(offset);
+  }
+
+  void insert_seg(byte *inode)
+  {
+    m_used_inodes[mach_read_from_4(inode + 4)].insert(
+      mach_read_from_2(inode + 8));
+  }
+
+  bool find(uint32_t page, uint16_t offset)
+  {
+    auto it= m_used_inodes.find(page);
+    if (it == m_used_inodes.end())
+      return false;
+    std::set<uint16_t> st= m_used_inodes[page];
+    return std::find(st.begin(), st.end(), offset) != st.end();
+  }
+
+  typename std::map<uint32_t, std::set<uint16_t>>::iterator begin()
+  { return m_used_inodes.begin(); }
+  typename std::map<uint32_t, std::set<uint16_t>>::iterator end()
+  { return m_used_inodes.end(); }
+};
+
+/* Get the used inode from the system tablespace
+@param used_inode list of inode information used in
+                  system tablespace
+@return error code */
+dberr_t fsp_get_sys_used_segment(inodeInfo *usedInodes)
+{
+ mtr_t mtr;
+ dberr_t err= DB_SUCCESS;
+ mtr.start();
+ /* Get TRX_SYS_FSEG_HEADER, TRX_SYS_DOUBLEWRITE_FSEG from
+ TRX_SYS_PAGE */
+ buf_block_t *block= buf_page_get(page_id_t(0, TRX_SYS_PAGE_NO),
+                                  0, RW_S_LATCH, &mtr);
+ if (!block)
+ {
+func_exit:
+   mtr.commit();
+   return err;
+ }
+
+ usedInodes->insert_seg(
+   block->page.frame + TRX_SYS + TRX_SYS_FSEG_HEADER);
+ usedInodes->insert_seg(
+   block->page.frame + TRX_SYS_DOUBLEWRITE + TRX_SYS_DOUBLEWRITE_FSEG);
+
+ /* Get rollback segment header page */
+ for (ulint rseg_id= 0; rseg_id < TRX_SYS_N_RSEGS; rseg_id++)
+ {
+   trx_rseg_t *rseg= &trx_sys.rseg_array[rseg_id];
+   if (rseg->space->id == 0)
+   {
+     block= buf_page_get_gen(rseg->page_id(), 0, RW_S_LATCH, nullptr,
+                             BUF_GET, &mtr, &err);
+     if (!block)
+       goto func_exit;
+     usedInodes->insert_seg(
+       block->page.frame + TRX_RSEG + TRX_RSEG_FSEG_HEADER);
+   }
+ }
+
+ block= buf_page_get_gen(page_id_t{0, DICT_HDR_PAGE_NO}, 0,
+                         RW_S_LATCH, nullptr, BUF_GET, &mtr, &err);
+ if (!block)
+   goto func_exit;
+ usedInodes->insert_seg(
+   block->page.frame + DICT_HDR + DICT_HDR_FSEG_HEADER);
+ block= buf_page_get_gen(page_id_t{0, FSP_IBUF_HEADER_PAGE_NO}, 0,
+                         RW_S_LATCH, nullptr, BUF_GET, &mtr, &err);
```

Comment:
Are we following the latching order here? It would be useful to add a comment that gives an example of a function that uses the same latching order.

If the server is in a kind of single-threaded mode (only this thread and some I/O handler threads) and if we are only reading here, then it could be better to use `buf_pool.page_fix()`; no need to hold multiple page latches to prevent concurrent modification.